### PR TITLE
add anchor generators used for detection

### LINKED ISSUE #3573: Anchor generator
**Is your feature request related to a problem? Please describe.**
Anchor generator for box detection is not in MONAI

**Describe the solution you'd like**
Add anchor generator

**Describe alternatives you've considered**
N/A

**Additional context**

### PR DESCRIPTION
**Pull request by Can-Zhao**
Signed-off-by: Can Zhao <canz@nvidia.com>

Fixes #3573 .

### Description
add anchor generators used for detection

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/detection/networks/utils/anchor_utils.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+
+        Returns:
+            For 2D images, returns [-w/2, -h/2, w/2, h/2];
+            For 3D images, returns [-w/2, -h/2, -d/2, w/2, h/2, d/2]
+        """
+        if device is None:
+            device = torch.device("cpu")
+        half_anchor_shapes = anchor_shapes / 2.0
+        base_anchors = torch.cat([-half_anchor_shapes, half_anchor_shapes], dim=1)
+        return base_anchors.round().to(dtype).to(device)
```
Comment (wyli) [inline monai/apps/detection/networks/utils/anchor_utils.py:326]: I came across this generator in detectron, looks like the non-integeters are handled differently https://github.com/facebookresearch/Detectron/blob/1809dd41c1ffc881c0d6b1c16ea38d08894f8b6d/detectron/modeling/generate_anchors.py#L113

**Review by wyli — COMMENTED** [review]
could you please add some unit tests for this module?

File: monai/apps/detection/networks/utils/anchor_utils.py (hunk 1 above)
Comment (Can-Zhao) [reply monai/apps/detection/networks/utils/anchor_utils.py:326]: Yes, they round anchor to int. I don't fully understand the rationale behind it. But we can also round it

Comment (wyli) [reply monai/apps/detection/networks/utils/anchor_utils.py:326]: > I might see your confusion. Detectron uses `TO_REMOVE=1`, So they have a lot of `+1` or `-1` in the code. Is that your concern? Or is `return base_anchors.round().to(dtype).to(device)` your concern? We could also choose to not round it

thanks, I'm not familiar with the details, just wanted to mention this difference, please feel free to ignore this one. 

(but there are a few autodetected issues please fix https://deepsource.io/gh/Project-MONAI/MONAI/run/0660e567-93c6-42b0-b965-16e5500c4937/python/)

**Comment by Can-Zhao** [comment]
> could you please add some unit tests for this module?

Thanks. Just added

File: monai/apps/detection/utils/anchor_utils.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+    def __init__(
+        self,
+        sizes: Sequence[Sequence[int]] = ((20, 30, 40),),
+        aspect_ratios: Sequence = (((0.5, 1), (1, 0.5)),),
+        indexing: str = "ij",
+    ) -> None:
+        super().__init__()
+
+        if not isinstance(sizes[0], Sequence):
+            # TODO change this
```
Comment (Nic-Ma) [inline monai/apps/detection/utils/anchor_utils.py:114]: What's the TODO mean here?

Thanks.

File: monai/apps/detection/utils/anchor_utils.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+            sizes = tuple((s,) for s in sizes)
+        if not isinstance(aspect_ratios[0], Sequence):
```
Comment (Nic-Ma) [inline monai/apps/detection/utils/anchor_utils.py:116]: Can we use `Sequence` to check data type?
Maybe better to use: https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/misc.py#L87

Thanks.

File: monai/apps/detection/utils/anchor_utils.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+                represents N anchor shapes for the current feature map.
+            dtype: target data type of the output Tensor.
+            device: target device to put the output Tensor data.
+
+        Returns:
+            For 2D images, returns [-w/2, -h/2, w/2, h/2];
+            For 3D images, returns [-w/2, -h/2, -d/2, w/2, h/2, d/2]
+        """
+        if device is None:
+            device = torch.device("cpu")
```
Comment (Nic-Ma) [inline monai/apps/detection/utils/anchor_utils.py:406]: Maybe we don't need to change the device if `device=None`.
Default to `cpu` is easy to cause errors.
You can refer to:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/inferers/utils.py#L98

Thanks.

**Review by Nic-Ma — APPROVED** [review]

File: monai/apps/detection/utils/anchor_utils.py (hunk 2 above)
Comment (Can-Zhao) [reply monai/apps/detection/utils/anchor_utils.py:114]: it comes from torchvision... I'm not sure. It seems they want to enhance the thid part

Comment (Nic-Ma) [reply monai/apps/detection/utils/anchor_utils.py:114]: Please remove it if not necessary?

Thanks.

File: monai/apps/detection/utils/anchor_utils.py [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+        anchors_over_all_feature_maps = self.grid_anchors(grid_sizes, strides)
+
+        anchors: List[List[torch.Tensor]] = []
+        for _ in range(batchsize):
+            anchors_in_image = list(anchors_over_all_feature_maps)
+            anchors.append(anchors_in_image)
+        return [torch.cat(anchors_per_image) for anchors_per_image in anchors]
+
+
+class AnchorGeneratorWithAnchorShape(AnchorGenerator, nn.Module):
```
Comment (wyli) [inline monai/apps/detection/utils/anchor_utils.py:320]: this nn.module could be removed, please help double-check (is it here because of torchscript?)
```suggestion
class AnchorGeneratorWithAnchorShape(AnchorGenerator):
```

File: monai/apps/detection/utils/anchor_utils.py [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+            dtype: target data type of the output Tensor.
+            device: target device to put the output Tensor data.
+
+        Returns:
+            For 2D images, returns [-w/2, -h/2, w/2, h/2];
+            For 3D images, returns [-w/2, -h/2, -d/2, w/2, h/2, d/2]
+        """
+        half_anchor_shapes = anchor_shapes / 2.0
+        base_anchors = torch.cat([-half_anchor_shapes, half_anchor_shapes], dim=1)
+        return base_anchors.round().to(dtype).to(device)
```
Comment (wyli) [inline monai/apps/detection/utils/anchor_utils.py:406]:
```suggestion
        return base_anchors.round().to(dtype=dtype, device=device)
```

File: tests/test_anchor_box.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+TEST_CASES_2D.append(
+    [
+        {"sizes": ((10, 12, 14, 16), (20, 24, 28, 32)), "aspect_ratios": ((1.0, 0.5, 2.0), (1.0, 0.5, 2.0))},
+        (5, 3, 128, 128),
+        ((5, 7, 64, 32), (5, 7, 32, 16)),
+    ]
+)
+
+
+class TestAnchorGenerator(unittest.TestCase):
```
Comment (wyli) [inline tests/test_anchor_box.py:35]: could you please help add a torchscript test? like this one:
https://github.com/Project-MONAI/MONAI/blob/1516ca758090dd8ed2890b64554a77f681a0e224/tests/test_se_blocks.py#L68-L72

**Review by wyli — COMMENTED** [review]
thanks, I put some comments inline.

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -57,6 +57,12 @@ repos:
               docs/source/conf.py
           )$
 
+  - repo: https://github.com/hadialqattan/pycln
```
Comment (wyli) [inline .pre-commit-config.yaml:60]: added `pycln` to auto-remove unused imports, cc @Project-MONAI/core-reviewers

File: monai/apps/detection/utils/anchor_utils.py [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+            the feature pyramid network (FPN).
+            Each elment of ``sizes`` is a Sequence which represents several anchor sizes for each feature map.
+        aspect_ratios: the aspect ratios of anchors. ``len(aspect_ratios) = len(sizes)``.
+            For 2D images, each element of ``aspect_ratios[i]`` is a Sequence of float.
+            For 3D images, each element of ``aspect_ratios[i]`` is a Sequence of 2 value Sequence.
+        indexing: choose from {'xy', 'ij'}, optional
+            Cartesian ('xy') or matrix ('ij', default) indexing of output.
+            Cartesian ('xy') indexing swaps axis 0 and 1, which is the setting inside torchvision.
+            matrix ('ij', default) indexing keeps the original axis not changed.
+            See also indexing in https://pytorch.org/docs/stable/generated/torch.meshgrid.html
```
Comment (mibaumgartner) [inline monai/apps/detection/utils/anchor_utils.py:83]: I find this a bit confusion, maybe the documentation could be improved here? The indexing of torch.meshgrid is always performed with 'ij' in the code and the axis are manually swapped. Does this also influence the aspect ratios and scales?

File: monai/apps/detection/utils/anchor_utils.py [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+
+            # compute anchor centers regarding to the image.
+            # shifts_centers is [x_center, y_center] or [x_center, y_center, z_center]
+            shifts_centers = [
+                torch.arange(0, size[axis], dtype=torch.int32, device=device) * stride[axis]
+                for axis in range(self.spatial_dims)
+            ]
+
+            # to support torchscript, cannot directly use torch.meshgrid(shifts_centers).
+            shifts_centers = list(torch.meshgrid(shifts_centers[: self.spatial_dims], indexing="ij"))
```
Comment (mibaumgartner) [inline monai/apps/detection/utils/anchor_utils.py:255]: Should this be refactored to the monai utility function?

File: monai/apps/detection/utils/anchor_utils.py [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+            [
+                torch.tensor(image_size[axis] // g[axis], dtype=torch.int64, device=device)
+                for axis in range(self.spatial_dims)
+            ]
+            for g in grid_sizes
+        ]
+
+        # Code below come from torchvision.models.detection.anchor_utils.AnchorGenerator.forward()
+        self.set_cell_anchors(dtype, device)
+        anchors_over_all_feature_maps = self.grid_anchors(grid_sizes, strides)
```
Comment (mibaumgartner) [inline monai/apps/detection/utils/anchor_utils.py:312]: Generally speaking for 3D: the patch size between batches is usually constant and re-generating the anchors every time might be a bit wasteful

**Review by mibaumgartner — COMMENTED** [review]
Left some small notes, the rest looks good to me.

File: monai/apps/detection/utils/anchor_utils.py (hunk 10 above)
Comment (wyli) [reply monai/apps/detection/utils/anchor_utils.py:255]: I spent some time on this one today, `monai.networks.utils.meshgrid_ij` is not compatible with torchscript...

Comment (Can-Zhao) [reply monai/apps/detection/utils/anchor_utils.py:255]: Yes, I also spent some time, and found out to support torchscript, you have to spell it out like `torch.meshgrid([shifts_centers[0], shifts_centers[1]], indexing="ij")` instead of torch.meshgrid(shifts_centers, indexing="ij")

File: monai/apps/detection/utils/anchor_utils.py (hunk 9 above)
Comment (Can-Zhao) [reply monai/apps/detection/utils/anchor_utils.py:83]: Yes, it is confusing. But that's how it works in torchvision. In the unit test, i tested its output and made sure it agrees with torchvision for 2D cases.

File: monai/apps/detection/utils/anchor_utils.py (hunk 11 above)
Comment (Can-Zhao) [reply monai/apps/detection/utils/anchor_utils.py:312]: I agree. We need to cache it

File: monai/apps/detection/utils/anchor_utils.py (hunk 10 above)
Comment (wyli) [reply monai/apps/detection/utils/anchor_utils.py:255]: removing the `indexing="ij"` will currently have the best compatibility across various versions. but it'll trigger a warning with the latest pytorch versions `torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument.`

File: monai/apps/detection/utils/anchor_utils.py (hunk 11 above)
Comment (Can-Zhao) [reply monai/apps/detection/utils/anchor_utils.py:314]: I still choose to output a list of same Tensor, I know it sounds stupid, but it is the output format in torchvision. I would like to make our anchor generator equivalent to torchvision for 2D cases...

File: monai/apps/detection/utils/anchor_utils.py (hunk 9 above)
Comment (Can-Zhao) [reply monai/apps/detection/utils/anchor_utils.py:83]: I added more docstring to explain our class with indexing 'xy' is equivalent to torchvision in 2D cases
